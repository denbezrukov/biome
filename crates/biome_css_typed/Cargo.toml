[package]
authors.workspace    = true
categories.workspace = true
description          = "Typed CSS property parser and analyzer for Biome"
edition.workspace    = true
homepage.workspace   = true
keywords.workspace   = true
license.workspace    = true
name                 = "biome_css_typed"
publish              = true
repository.workspace = true
version              = "0.1.0"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
biome_css_syntax  = { workspace = true }
biome_diagnostics = { workspace = true }

[lints]
workspace = true
