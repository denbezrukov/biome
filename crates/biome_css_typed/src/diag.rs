use biome_diagnostics::{Diagnostic, Location, Severity, console::fmt::Formatter};
use biome_text_size::TextRange;
use std::fmt;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CssPropertyDiagnostic {
    pub kind: CssPropertyDiagnosticKind,
    pub range: Option<TextRange>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CssPropertyDiagnosticKind {
    UnknownProperty,
    InvalidValue,
    UnexpectedToken,
    DuplicateComponent,
}

impl CssPropertyDiagnosticKind {
    fn message(self) -> &'static str {
        match self {
            CssPropertyDiagnosticKind::UnknownProperty => "Unknown CSS property.",
            CssPropertyDiagnosticKind::InvalidValue => "Invalid value for CSS property.",
            CssPropertyDiagnosticKind::UnexpectedToken => "Unexpected token in CSS property value.",
            CssPropertyDiagnosticKind::DuplicateComponent => {
                "Duplicate component in CSS property value."
            }
        }
    }
}

impl Diagnostic for CssPropertyDiagnostic {
    fn severity(&self) -> Severity {
        Severity::Error
    }

    fn message(&self, fmt: &mut Formatter<'_>) -> std::io::Result<()> {
        fmt.write_str(self.kind.message())
    }

    fn description(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt.write_str(self.kind.message())
    }

    fn location(&self) -> Location<'_> {
        let mut builder = Location::builder();
        if let Some(range) = &self.range {
            builder = builder.span(range);
        }
        builder.build()
    }
}
