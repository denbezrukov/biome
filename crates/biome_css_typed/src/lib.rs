//! Typed CSS property parser and analyzer for Biome.
//!
//! This crate builds on top of `biome_css_syntax` and provides
//! a strongly-typed representation of CSS property values for use
//! in analyzers and lint rules.

pub mod api;
pub mod diag;
pub mod parse;
pub mod props;
pub mod util;
