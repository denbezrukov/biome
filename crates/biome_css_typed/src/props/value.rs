use super::{padding::PaddingShorthand, visibility::Visibility};

#[derive(Debug, Clone, PartialEq)]
pub enum PropertyValue {
    Visibility(Visibility),
    PaddingShorthand(PaddingShorthand),
    // Display(super::display::Display),  // later
}
