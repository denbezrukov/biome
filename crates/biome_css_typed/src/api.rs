use biome_css_syntax::CssGenericProperty;

use crate::diag::CssPropertyDiagnostic;
use crate::props::{dispatch, value::PropertyValue};

pub fn parse_property_typed(
    property: &CssGenericProperty,
) -> Result<PropertyValue, CssPropertyDiagnostic> {
    dispatch::parse_property(property)
}
