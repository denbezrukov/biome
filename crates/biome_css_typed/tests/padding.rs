#[path = "utils.rs"]
mod utils;

use biome_css_typed::api;
use biome_css_typed::diag::CssPropertyDiagnostic;
use biome_css_typed::props::value::TypedPropertyValue;
use utils::parse_generic_property;

fn parse_padding(source: &str) -> Result<TypedPropertyValue, CssPropertyDiagnostic> {
    let property = parse_generic_property("padding", source);
    api::parse_property_typed(&property)
}

#[test]
fn padding_valid_values() {
    let results: Vec<_> = ["10px", "10px 20%", "1em 2em 3em", "1px 2px 3px 4px"]
        .into_iter()
        .map(parse_padding)
        .collect();
    insta::assert_debug_snapshot!("padding_valid_values", results);
}

#[test]
fn padding_invalid_values() {
    let results: Vec<_> = ["-1px", "1px junk", "inherit"]
        .into_iter()
        .map(parse_padding)
        .collect();
    insta::assert_debug_snapshot!("padding_invalid_values", results);
}
