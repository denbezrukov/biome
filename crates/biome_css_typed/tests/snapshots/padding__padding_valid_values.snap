---
source: crates/biome_css_typed/tests/padding.rs
expression: results
---
[
    Ok(
        PaddingShorthand(
            PaddingShorthand(
                Box4 {
                    top: Length(
                        Length {
                            value: 10.0,
                            unit: Px,
                        },
                    ),
                    right: None,
                    bottom: None,
                    left: None,
                },
            ),
        ),
    ),
    Ok(
        PaddingShorthand(
            PaddingShorthand(
                Box4 {
                    top: Length(
                        Length {
                            value: 10.0,
                            unit: Px,
                        },
                    ),
                    right: Some(
                        Percentage(
                            Percentage(
                                20.0,
                            ),
                        ),
                    ),
                    bottom: None,
                    left: None,
                },
            ),
        ),
    ),
    Ok(
        PaddingShorthand(
            PaddingShorthand(
                Box4 {
                    top: Length(
                        Length {
                            value: 1.0,
                            unit: Em,
                        },
                    ),
                    right: Some(
                        Length(
                            Length {
                                value: 2.0,
                                unit: Em,
                            },
                        ),
                    ),
                    bottom: Some(
                        Length(
                            Length {
                                value: 3.0,
                                unit: Em,
                            },
                        ),
                    ),
                    left: None,
                },
            ),
        ),
    ),
    Ok(
        PaddingShorthand(
            PaddingShorthand(
                Box4 {
                    top: Length(
                        Length {
                            value: 1.0,
                            unit: Px,
                        },
                    ),
                    right: Some(
                        Length(
                            Length {
                                value: 2.0,
                                unit: Px,
                            },
                        ),
                    ),
                    bottom: Some(
                        Length(
                            Length {
                                value: 3.0,
                                unit: Px,
                            },
                        ),
                    ),
                    left: Some(
                        Length(
                            Length {
                                value: 4.0,
                                unit: Px,
                            },
                        ),
                    ),
                },
            ),
        ),
    ),
]
